File: SmillaAI/SenseLED.py
#!/usr/bin/env python3
########################################################################
# Filename    : SenseLED.py
# Description : Controlling an led by infrared Motion sensor.
# Author      : freenove
# modification: 2018/08/03
########################################################################
import RPi.GPIO as GPIO
from time import sleep

ledPin = 12    # define the ledPin
sensorPin = 11    # define the sensorPin

def setup():
	print ('Program is starting...')
	GPIO.setmode(GPIO.BOARD)       # Numbers GPIOs by physical location
	GPIO.setup(ledPin, GPIO.OUT)   # Set ledPin's mode is output
	GPIO.setup(sensorPin, GPIO.IN)    # Set sensorPin's mode is input

def irsense_on_callback(channel):
	if GPIO.input(sensorPin)==GPIO.HIGH:
		print('Rise!')
		GPIO.output(ledPin,GPIO.HIGH)
	else:
		print('Fall')
		GPIO.output(ledPin,GPIO.LOW)

def loop():
	GPIO.output(ledPin,GPIO.LOW)
	sleep(5)
	GPIO.add_event_detect(sensorPin, GPIO.BOTH)
	GPIO.add_event_callback(sensorPin, callback=irsense_on_callback)
	sleep(100000)

def destroy():
	GPIO.remove_event_detect(sensorPin)
	GPIO.cleanup()                     # Release resource

if __name__ == '__main__':     # Program start from here
	setup()
	try:
		loop()
	except KeyboardInterrupt:  # When 'Ctrl+C' is pressed, the child program destroy() will be  executed.
		destroy()

File: SmillaAI/wiring/SenseLED.c
/**********************************************************************
* Filename    : SenseLED.c
* Description : Controlling an led by infrared Motion sensor.
* Author      : freenove
* modification: 2016/06/12
**********************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <wiringPi.h>

#define LED_PIN    1  	//define the ledPin
#define SENSOR_PIN 0		//define the sensorPin

char* getTimeString(void)
{
	time_t rawtime;
	struct tm * timeinfo;

	time( &rawtime );
	timeinfo = localtime( &rawtime );
	return asctime( timeinfo );
}

int loop(void)
{
	pinMode(LED_PIN, OUTPUT); 
	pinMode(SENSOR_PIN, INPUT);
	int i = 0;	
	int status = LOW;
	while (1) {		
		if (digitalRead(SENSOR_PIN) == HIGH){ //if read sensor for high level
			if (HIGH != status) {
				digitalWrite(LED_PIN, HIGH);   //led on
				printf("%s %i: led on...\n", getTimeString(), i++);
				status = HIGH;
			}
		}
		else {				
			if (LOW != status)
			{
				digitalWrite(LED_PIN, LOW);   //led off
				printf("%s %i: ...led off\n", getTimeString(), i++);
				status = LOW;
			}
		}
	}

	return 0;
}

void handler(void)
{
	system("kill -USR1 $(pgrep raspivid)");
	delay(5000);
	system("kill -USR1 $(pgrep raspivid)");

	//printf("Capture status: %i\n", system("kill -USR1 $(pgrep raspistill)"));
	//printf("Sensor status: %i\n", digitalRead(SENSOR_PIN));
}

int wait_for_event(void)
{
	int ec = wiringPiISR(SENSOR_PIN, INT_EDGE_RISING, &handler);
	if (ec != -0) { 
		printf("Setup of handler failed!\n");
		return 1; 
	}

	while(1) {
		delay(10000);
		// printf(".\n");
	}

	return 0;
}

int main(void)
{
	printf("Starting...\n");

	if(wiringPiSetup() == -1){ //when initialize wiring failed,print messageto screen
		printf("Setup of wiringPi failed!\n");
		return 1; 
	}
	
	// return loop();
	return wait_for_event();
}

